symengine/integer.cpp

```diff
@@ -10,10 +10,9 @@ Integer::Integer(int i)
 
 std::size_t Integer::__hash__() const
 {
-    std::hash<long long int> hash_fn;
     // only the least significant bits that fit into "signed long int" are
     // hashed:
-    return hash_fn(this->i.get_si());
+    return this->i.get_ui() * mpz_sgn(this->i.get_mpz_t());
```

Comment:
For ints with small absolute values hashes are the same in all python versions (https://docs.python.org/3/library/stdtypes.html#hashing-of-numeric-types), so that tests like `set((2*x**2*y).args) == set((2, x**2, y))` can be kept as it is.

I'll look into whether speed is a concern. I'll revert back to 08ecc0b123548e9e39f74f39383b8c89b78e4352, get this merged in and then see what should be done.